ext/FluxMooncakeExt.jl

```diff
@@ -0,0 +1,173 @@
+
+"""
+    Flux.gradient(f, args::Moonduo...)
+
+This uses Mooncake.jl to compute the derivative,
+which is both stored within `Moonduo` and returned.
+Similar to the Enzyme.jl methods like `Flux.gradient(f, m::Duplicated)`.
+
+# Example
+
+```julia
+julia> using Flux
+
+julia> model = Chain(Dense([3.0;;]));
+
+julia> Flux.gradient(model, [1]) do m, x  # computed using Zygote
+         sum(abs2, m(x))
+       end
+((layers = ((weight = [6.0;;], bias = [6.0], σ = nothing),),), [18.0])
+
+julia> using Fluxperimental, Mooncake
+
+julia> dup_model = Moonduo(model);  # allocates space for gradient
+
+julia> Flux.gradient(dup_model, Moonduo([1])) do m, x  # Mooncake, returns the same
+         sum(abs2, m(x))
+       end
+((layers = ((weight = [6.0;;], bias = [6.0], σ = nothing),),), nothing)
+
+julia> dup_model  # same gradient is also stored within Duplicated
+Moonduo(
+  Chain(
+    Dense(1 => 1),                      # 2 parameters
+  ),
+  # norm(∇) ≈ 8.49
+)
+
+julia> Flux.destructure((weight = [6.0;;], bias = [6.0]))[1] |> norm
+8.48528137423857
+
+julia> Flux.gradient(dup_model, Moonduo([1]); zero=false) do m, x  # grad accumulation
+         sum(abs2, m(x))
+       end
+((layers = ((weight = [12.0;;], bias = [12.0], σ = nothing),),), nothing)
+```
+
+!!! note
+    At present there is no way to mark some arguments constant.
+    Instead of `gradient(loss, Duplicated(model), Const(data))`,
+    you can write `gradient(m -> loss(m, data), Moonduo(model))`.
+"""
+Flux.gradient(f, args::Moonduo...; zero::Bool=true) = _moon_withgradient(f, args...; zero).grad
+
+"""
+    Flux.withgradient(f, args::Moonduo...)
+
+This should return the same answer as `withgradient(f, model, args...)`,
+but it uses Mooncake.jl instead of Zygote.jl to compute the derivative.
+
+# Example
+
+```julia
+julia> using Flux, Fluxperimental, Mooncake
+
+julia> model = Chain(Embedding([1.1 2.2 3.3]), Dense([4.4;;]), only);
+
+julia> model(3)
+14.52
+
+julia> Flux.withgradient(m -> m(3), model)  # this uses Zygote
+(val = 14.52, grad = ((layers = ((weight = [0.0 0.0 4.4],), (weight = [3.3;;], bias = [1.0], σ = nothing), nothing),),))
+
+julia> Flux.withgradient(m -> m(3), Moonduo(model))  # this uses Mooncake
+(val = 14.52, grad = ((layers = ((weight = [0.0 0.0 4.4],), (weight = [3.3;;], bias = [1.0], σ = nothing), nothing),),))
+```
+
+!!! warning
+    With Zygote, the function `f` may return Tuple or NamedTuple, with the loss as the first element.
+    This feature is not supported here, for now.
+"""
+Flux.withgradient(f, args::Moonduo...; zero::Bool=true) = _moon_withgradient(f, args...; zero)
+
+function _moon_withgradient(f, args::Moonduo...; zero)
+  plain = map(x -> x.val, args)
+  rule = Mooncake.build_rrule(f, plain...)
+
+  for x in args
+    _check_mutable(x)
+    zero && Mooncake.set_to_zero!!(x.dval)
+  end
+  coduals = map(x -> Mooncake.CoDual(x.val, x.dval), args)
+  val, _ = Mooncake.__value_and_gradient!!(rule, Mooncake.zero_codual(f), coduals...)
```

Comment:
Is there any chance at all that `f` will contain trainable parameters, or does Flux insist that you not do that?

Reply:
Yea `Flux.gradient(f, x, y)` has length 2, alla Zygote. I agree that's not the maximally flexible thing, and very occasionally you end up with `gradient(|>, x, f)`... but in real use it seems like never.

Reply:
Cool